GPIO resistor setting has no effect

### Describe the bug
Can neither set pullup nor pulldown works on raspberry pi (tested pi3b and pi4b).
Pulldown was not implemented as it seems - so that is okay.
My workaround is to use resistor: 'off' and change desired behavior for all used pins in /boot/config.txt

### Environment
- MQTTany most recent branch add-rpi4
- Platform: Raspberry Pi 3B
- Platform: Raspberry Pi 4B

### Configuration

```yaml
gpio:
 my-pin:
    pin: '17'
    resistor: 'pullup'
    digital:
     interrupt: 'falling'
```

### Logs
## I added some debug outputs

```log
2021-01-10 17:52:52,003 [DEBUG] [gpio    ] [core.gpio.pin.digital   ] Line is 17
2021-01-10 17:52:52,004 [DEBUG] [gpio    ] [core.gpio.pin.digital   ] Mode is PinMode.INPUT
2021-01-10 17:52:52,006 [DEBUG] [gpio    ] [core.gpio.pin.digital   ] Edge is PinEdge.FALLING
2021-01-10 17:52:52,007 [DEBUG] [gpio    ] [core.gpio.pin.digital   ] Bias is PinBias.PULL_UP
2021-01-10 17:52:52,009 [ERROR] [gpio    ] [core.gpio.pin.digital   ] Failed to setup GPIO17 using cdev, an invalid argument was given
```

Let me first say, that your piece of software is really helpful. Thanks a lot for that. I think I may contribute to it a little.

Let me answer your question:
The behaviour on both devices is identical.

The Line
[ERROR] [gpio ] [core.gpio.pin.digital ] Failed to setup GPIO17 using cdev, an invalid argument was given
is the error.

I just printed out the arguments that are passed to create the GPIO for debug purposed to make sure they are valid (but they are). Of course I use the -vv option.

Why do you think pulldowns are not supported on RPI devices? You can set almost all I/Os (except those used for I2C) to Pulldown. I did this modification for IO4 IO5 and IO6 in the config.txt file as a workaround.

```
gpio=4-6=ip
gpio=4-6=pd
```

Also there may be issues with interrupts and debouncing. I tried to follow your code in mqttany/gpio/pins/digital.py, but did not yet understand it 100%.

Shouldn't it look more like this.
https://raspberrypi.stackexchange.com/a/76738

In my application, I need to switch on and dim lightbulbs with these button. Therefore, I would like to receive a message through MQTT (check for an event on a rising edge if the value is still high after that period - debouncing, then send "ON"). Of course no other events are valid in this period. I also need to detect the falling edge afterwards, because I would like to start the process of dimming the lightbulb in openhab until the button is released (falling edge send to openhab "OFF").

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,8 +2,14 @@
 
 ## Development
 
+* **Added**
+  * Kernel version check to determine GPIO cdev availability and capabilities.
+
 * **Fixed**
   * *Unknown* and *Generic Linux* boards missing superclass init.
+  * GPIO error `Failed to setup GPIOxx using cdev, an invalid argument was given`
+    caused by attempting to use cdev bias setting that was only introduced in Linux
+    kernel v5.5. Fixes [#106](https://github.com/CrazyIvan359/mqttany/issues/106).
 
 ## 0.13.0
 

--- a/mqttany/gpio/__init__.py
+++ b/mqttany/gpio/__init__.py
@@ -37,6 +37,7 @@ __all__ = [
 ]
 
 import os
+import platform
 import adafruit_platformdetect
 from adafruit_platformdetect.constants import boards as board_ids
 
@@ -45,9 +46,19 @@ from .common import log, Mode, PinMode, PinBias, PinEdge, PinAlternate
 from .boards import get_board, Unknown
 
 board = get_board()
-common.cdev = os.access(
-    f"/dev/gpiochip{board.chips[0]}", os.F_OK | os.R_OK | os.W_OK, effective_ids=True
-)
+
+# Check cdev GPIO availability and capabilities
+kernel_version = tuple([int(s) for s in platform.release().split(".")[:2]])
+if kernel_version >= (4, 9):
+    common.cdev = os.access(
+        f"/dev/gpiochip{board.chips[0]}",
+        os.F_OK | os.R_OK | os.W_OK,
+        effective_ids=True,
+    )
+if kernel_version >= (5, 5):
+    common.cdev_bias = True
+
+# Check sysfs GPIO availability
 common.sysfs = os.access(
     "/sys/class/gpio/export", os.F_OK | os.W_OK, effective_ids=True
 )
@@ -77,30 +88,18 @@ def init() -> None:
             log.warn("Attempting to fall back to sysfs")
             if common.sysfs:
                 log.debug("Falling back to sysfs")
-            elif os.access("/sys/class/gpio/export", os.F_OK, effective_ids=True):
-                log.warn(
-                    "Detected sysfs GPIO but this account (%s) does not have R/W "
-                    "permissions",
-                    os.getlogin(),
-                )
-                log.warn(
-                    "Read/Write access is required on the following directory: "
-                    "'/sys/class/gpio/export'"
-                )
 
-        else:
-            if common.sysfs:
-                log.debug("Detected sysfs GPIO interface")
-            elif os.access("/sys/class/gpio/export", os.F_OK, effective_ids=True):
-                log.warn(
-                    "Detected sysfs GPIO but this account (%s) does not have R/W "
-                    "permissions",
-                    os.getlogin(),
-                )
-                log.warn(
-                    "Read/Write access is required on the following directory: "
-                    "'/sys/class/gpio/export'"
-                )
+        if common.sysfs:
+            log.debug("Detected sysfs GPIO interface")
+        elif os.access("/sys/class/gpio/export", os.F_OK, effective_ids=True):
+            log.warn(
+                "Detected sysfs GPIO but this account (%s) does not have "
+                "permission to access '/sys/class/gpio/export'",
+                os.getlogin(),
+            )
+            log.warn(
+                "Access is required on the following directory: " "'/sys/class/gpio/'"
+            )
 
         if not common.cdev and not common.sysfs:
             log.warn("No GPIO interface is available, some features may not work")

--- a/mqttany/gpio/common.py
+++ b/mqttany/gpio/common.py
@@ -42,6 +42,7 @@ import logger
 
 log = logger.get_logger("core.gpio")
 cdev = False
+cdev_bias = False
 sysfs = False
 
 

--- a/mqttany/gpio/pins/digital.py
+++ b/mqttany/gpio/pins/digital.py
@@ -116,16 +116,27 @@ class Digital(Pin):
     def setup(self) -> bool:
         """Setup the pin before use, returns ``True`` on success."""
         if not self._setup:
+            sysfs_failover = True
             if common.cdev:
+                if self._bias != PinBias.NONE and not common.cdev_bias:
+                    self._bias = PinBias.NONE
+                    log.warn(
+                        "Unable to set resistor on %s, this requires a kernel version "
+                        "of 5.5 or higher",
+                        self.name,
+                    )
                 if self._setup_cdev():
                     if self.edge:
                         self._interrupt_handler = cdevInterrupt(self)
                     self._setup = True
-                elif common.sysfs and self._setup_sysfs():  # failover to sysfs
-                    if self.edge:
-                        self._interrupt_handler = sysfsInterrupt(self)
-                    self._setup = True
-            elif common.sysfs:
+                    sysfs_failover = False
+            if common.sysfs and sysfs_failover:
+                if self._bias != PinBias.NONE:
+                    self._bias = PinBias.NONE
+                    log.warn(
+                        "Unable to set resistor on %s as sysfs does not support this",
+                        self.name,
+                    )
                 if self._setup_sysfs():
                     if self.edge:
                         self._interrupt_handler = sysfsInterrupt(self)
@@ -135,6 +146,13 @@ class Digital(Pin):
 
         return self._setup
 
+    def _set_bias_pinctrl(self):
+        """
+        For cdev on kernel <5.5 and sysfs, attempts to set bias resistor using `ioctl`
+        calls to `pinctrl`
+        """
+        pass
+
     def _setup_cdev(self) -> bool:
         try:
             log.debug("Setting up pin %s using chardev", self.name)
@@ -142,7 +160,7 @@ class Digital(Pin):
                 f"/dev/gpiochip{self.chip}",
                 self.line,
                 periphery_PinMode[self.mode],
-                edge=periphery_PinEdge[self.edge],
+                edge="none" if not common.cdev_bias else periphery_PinEdge[self.edge],
                 bias=periphery_PinBias[self.bias],
                 label="MQTTany",
             )

